### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by mohavro**
## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test addition/update
- [ ] Dependency update

## Related Issues 

Closes #

## Changes Made

-Addition of 3 codebase / knowledgebase .md files 
- tech_spec.md
- DashFin-financial-asset-relationship-db-wiki-v2.1.md
- codebase_context.md

## Testing

- [X] Unit tests pass locally
- [X] Integration tests pass locally
- [ ] Manual testing completed
- [ ] New tests added for new functionality

### Test Commands

```bash
# Commands used to test
pytest
npm test
```

### Code Quality

- [X] My code follows the project's style guidelines
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### Documentation

- [ ] I have updated the README.md (if needed)
- [ ] I have updated the CHANGELOG.md
- [X] I have updated relevant documentation files
- [ ] I have added/updated docstrings

### Dependencies

- [X] I have checked for dependency conflicts
- [ ] I have updated requirements.txt (if needed)
- [ ] I have updated package.json (if needed)

### Branch Management

- [X] My branch is up to date with main
- [X] I will delete this branch after merge
- [ ] This branch has a descriptive name
- [X] This PR has been open for less than 2 weeks

## Additional Notes

**For Reviewers**: Please ensure the branch is deleted after merging this PR to maintain repository hygiene.

**Related Documentation**:
- [Contributing Guidelines](../CONTRIBUTING.md)
- [Branch Cleanup Guidelines](../BRANCH_CLEANUP_QUICK_REFERENCE.md)

---
## Summary by cubic
Adds comprehensive documentation to clarify architecture, APIs, and data/visualization flow, improving onboarding and contributor guidance. Adds the DashFin financial-asset DB wiki v2.1 and a Codebase Overview, updates the Quick Test Reference, and expands the test suite with new unit/integration coverage for auth, database, MCP server, real data fetcher, visuals, reports, sample data, and workflow/security validators.

- **Bug Fixes**
  - Gradio: show_formula_details now returns a default (empty figure, hidden error) tuple to ensure consistent outputs and avoid UI errors.

- **Refactors**
  - Added docstrings and small code cleanups across core modules to improve clarity, consistent returns, and staticmethod usage.

<sup>Written for commit f2bc08c1b3154971ca917d008336659f131140ac. Summary will update on new commits. <a href="https://cubic.dev/pr/DashFin-FarDb/financial-asset-relationship-db/pull/648">Review in cubic</a></sup>

### **Auto-created Ticket**

[#649](https://github.com/DashFin-FarDb/financial-asset-relationship-db/issues/649)

### **Description**
- Added three new documentation files to enhance the codebase knowledgebase.
- Introduced `codebase_context (2).md` for a detailed overview of the repository's architecture and conventions.
- Updated `tech_spec.md` to improve technical specifications.
- Added `DashFin-financial-asset-relationship-
...(truncated)

___

## **CodeAnt-AI Description**
Add comprehensive unit and integration tests, stricter PR-agent config checks, and clearer UI/analysis outputs

### What Changed
- Added many new unit tests (auth, mcp_server, graph visuals, db models, server concurrency, and more) that increase coverage and validate CLI, JWT auth, graph thread-safety, 3D layout output, and edge cases.
- Strengthened integration checks for .github/pr-agent-config.yml and related workflows: detect hardcoded credentials, enforce 2-space indentation, ban specific legacy keys (fallback, chunk_size, max_tokens misuse), and ensure workflow files do not reference deleted or removed patterns.
- Improved requirements and documentation tests and parsers: preserve package name casing, normalize version specifiers, and validate sensible version constraints for critical/test packages and type stubs.
- Updated application UI/analysis behavior descriptions and return values: metrics text and formulaic analysis now produce clearer, human-readable summaries and defined return structures; asset info and refresh functions return explicit empty/placeholder outputs when assets are missing.
- Clarified and hardened secret-detection heuristics: multiple heuristics (prefixes, long/high-entropy strings, inline credentials, and sensitive key checks) now fail tests when potential secrets or unsafe placeholders are present in PR agent config.

### Impact
`✅ More test coverage across authentication, graph visuals, and server components`
`✅ Clearer network metrics and formula analysis summaries in the UI`
`✅ Fewer unsafe PR configs by detecting hardcoded credentials and invalid workflow references`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: tests/unit/test_schema_report.py [hunk 1]
```diff
@@ -457,3 +459,144 @@ def test_different_output_for_modified_graph(populated_graph):
...
+        if metrics["relationship_density"] <= 10:
+            assert (
+                "sparse" in report.lower()
+                or "adding more relationships" in report.lower()
+            )
+
+    @staticmethod
+    def test_schema_report_all_relationship_types():
+        """Test that report includes all relationship types."""
+        graph = create_sample_database()
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_schema_report.py]: `from src.logic.asset_graph import AssetRelationshipGraph
from src.models.financial_models import AssetClass, Equity
from src.data.sample_data import create_sample_database`
```suggestion
    def test_schema_report_empty_graph():
        """Test generating schema report for empty graph."""
        graph = AssetRelationshipGraph()

        report = generate_schema_report(graph)

        assert report is not None
        assert "Total Assets: 0" in report
        assert "Total Relationships: 0" in report

    @staticmethod
    def test_schema_report_single_asset():
        """Test schema report with only one asset."""
        graph = AssetRelationshipGraph()
        equity = Equity(
            id="SINGLE",
            symbol="SGL",
            name="Single Asset",
            asset_class=AssetClass.EQUITY,
            sector="Tech",
            price=100.0,
        )
        graph.add_asset(equity)

        report = generate_schema_report(graph)

        assert "Total Assets: 1" in report
        assert "equity: 1" in report.lower()

    @staticmethod
    def test_schema_report_high_density():
        """Test schema report with high relationship density."""
        graph = AssetRelationshipGraph()

        # Add multiple assets in same sector for high connectivity
        for i in range(10):
            equity = Equity(
                id=f"ASSET{i}",
                symbol=f"A{i}",
                name=f"Asset {i}",
                asset_class=AssetClass.EQUITY,
                sector="Technology",  # Same sector
                price=100.0 + i,
            )
            graph.add_asset(equity)

        graph.build_relationships()

        report = generate_schema_report(graph)

        # Should recommend normalization for high density
        metrics = graph.calculate_metrics()
        if metrics["relationship_density"] > 30:
            assert "normalization" in report.lower()

    @staticmethod
    def test_schema_report_low_density():
        """Test schema report with low relationship density."""
        graph = AssetRelationshipGraph()

        # Add assets in different sectors for low connectivity
        for i in range(5):
            equity = Equity(
                id=f"ASSET{i}",
                symbol=f"A{i}",
                name=f"Asset {i}",
                asset_class=AssetClass.EQUITY,
                sector=f"Sector{i}",  # Different sectors
                price=100.0 + i,
            )
            graph.add_asset(equity)

        graph.build_relationships()

        report = generate_schema_report(graph)

        # Should recommend adding relationships for low density
        metrics = graph.calculate_metrics()
        if metrics["relationship_density"] <= 10:
            assert (
                "sparse" in report.lower()
                or "adding more relationships" in report.lower()
            )

    @staticmethod
    def test_schema_report_all_relationship_types():
        """Test that report includes all relationship types."""
        graph = create_sample_database()
```

File: tests/unit/test_mcp_server.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+                    sector=sector,
+                    price=price,
+                )
+                return f"Successfully validated: {new_equity.name}"
+            except (ValueError, TypeError) as e:
+                return f"Validation Error: {str(e)}"
+
+        result = add_equity_node("", "TEST", "Test", "Tech", 100.0)
+
+        assert "Validation Error" in result
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_mcp_server.py]: `assert "Validation Error" in result`
```suggestion
        result = add_equity_node("", "TEST", "Test", "Tech", 100.0)

        assert "Validation Error" in result
```

**Review by mohavro — COMMENTED** [review]
Do not comment on problems not within the scope of the PR. No new problems unless breaking code that needs urgent change.

File: tests/unit/test_real_data_fetcher.py [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+    def test_create_database_uses_cache_when_available(mock_load):
+        """Test that create_database uses cache when available."""
+        mock_graph = AssetRelationshipGraph()
+        mock_load.return_value = mock_graph
+
+        with tempfile.NamedTemporaryFile(suffix=".pkl") as tmp:
+            tmp.write(b"dummy")
+            tmp.flush()
+
+            fetcher = RealDataFetcher(use_cache=True, cache_path=Path(tmp.name))
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_real_data_fetcher.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/648#discussion_r2705416508)

File: tests/unit/test_mcp_server.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+        ]
+
+        for t in threads:
+            t.start()
+        for t in threads:
+            t.join()
+
+        # All threads should complete successfully
+        assert len(results) == 3
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_mcp_server.py]: `expected_ids = {f"TEST{i}" for i in range(15)}
assets = wrapped.assets
assert set(assets.keys()) == expected_ids`
```suggestion
    def test_concurrent_access_is_safe():
        """Test that concurrent access to graph is thread-safe."""
        graph = AssetRelationshipGraph()
        lock = threading.Lock()
        wrapped = mcp_server._ThreadSafeGraph(graph, lock)

        results = []

        def add_assets(start_id: int):
            for i in range(5):
                equity = Equity(
                    id=f"TEST{start_id + i}",
                    symbol=f"T{start_id + i}",
                    name=f"Test {start_id + i}",
                    asset_class=AssetClass.EQUITY,
                    sector="Tech",
                    price=100.0 + i,
                )
                wrapped.add_asset(equity)
            results.append(True)

        threads = [
            threading.Thread(target=add_assets, args=(i * 10,)) for i in range(3)
        ]

        for t in threads:
            t.start()
        for t in threads:
            t.join()

        # All threads should complete successfully
        assert len(results) == 3

```

File: tests/unit/test_mcp_server.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+class TestConcurrency:
+    """Test thread safety of the MCP server."""
+
+    @staticmethod
+    def test_graph_lock_exists():
+        """Test that global graph lock exists."""
+        assert hasattr(mcp_server, "_graph_lock")
+        assert isinstance(mcp_server._graph_lock, threading.Lock)
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_mcp_server.py]: `assert isinstance(mcp_server._graph_lock, type(threading.Lock()))`
```suggestion
    def test_graph_lock_exists():
        """Test that global graph lock exists."""
        assert hasattr(mcp_server, "_graph_lock")
        assert isinstance(mcp_server._graph_lock, threading.Lock)

```

File: tests/unit/test_mcp_server.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+        ]
+
+        for t in threads:
+            t.start()
+        for t in threads:
+            t.join()
+
+        # All threads should complete successfully
+        assert len(results) == 3
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_mcp_server.py]: `expected_ids = {f"TEST{i}" for i in range(15)}
assets = wrapped.assets
assert set(assets.keys()) == expected_ids`
```suggestion
    def test_concurrent_access_is_safe():
        """Test that concurrent access to graph is thread-safe."""
        graph = AssetRelationshipGraph()
        lock = threading.Lock()
        wrapped = mcp_server._ThreadSafeGraph(graph, lock)

        results = []

        def add_assets(start_id: int):
            for i in range(5):
                equity = Equity(
                    id=f"TEST{start_id + i}",
                    symbol=f"T{start_id + i}",
                    name=f"Test {start_id + i}",
                    asset_class=AssetClass.EQUITY,
                    sector="Tech",
                    price=100.0 + i,
                )
                wrapped.add_asset(equity)
            results.append(True)

        threads = [
            threading.Thread(target=add_assets, args=(i * 10,)) for i in range(3)
        ]

        for t in threads:
            t.start()
        for t in threads:
            t.join()

        # All threads should complete successfully
        assert len(results) == 3

```

File: codebase_context (2).md [hunk 7]
```diff
@@ -0,0 +1,99 @@
+# Codebase Overview & Contributor Reference
```
[the remark this answers is not in the document]
Comment (mohavro) [reply codebase_context (2).md]: @coderabbitai suggest the code required

**Review by mohavro — COMMENTED** [review]
merge PR

**Review by mohavro — COMMENTED** [review]
no new comments unless they are blocking or critical issues. Merge PR when all tasks completed.
